### ISSUE TITLE
History copying optimization and/or UX improvement

### ISSUE THREAD

**Issue by carlfeberhard** [galaxyproject/galaxy#2804]
> if you have a sufficiently large collection (in my case 300+ datasets) it takes &gt;10 min to copy it into a new history. crazy

edit:
Related to: #2526 

> [nekrut] but process continues in the background, so eventially it does get copied, but why soooooooooo long?

**Comment by carlfeberhard** [comment]
@natefoo (objectStore peoples:) @afgane @dannon Is there a reason I'm missing for needing  `hda.set_size()` here: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/model/__init__.py#L2326

If this is a copy then it's re-using the dataset - which has a calculated size already, right? I wanted to check if there's any exotic use case you guys might know of that requires it.

**Comment by natefoo** [comment]
It does seem unnecessary, but is it actually setting the size? If `dataset.file_size` is not null, this should skip actually doing the `stat()`s and calculation: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/model/__init__.py#L1760

**Comment by jmchilton** [comment]
Just a heads up on this issue also - @blankenberg theorized that this operation may result in re-indexing bam files and Anton had a huge collection of bam files as part of this copy.

**Comment by carlfeberhard** [comment]
@natefoo yeah - it's skipping it but there's a query to pull the dataset in order to see if it can skip it. I might end up having to join the dataset anyway, but in any event if we don't need it I'll take it out.

@jmchilton @blankenberg I tried a simple bam copy (1.bam, has a bai, same user, same history) and `datatypes.Bam.set_meta` wasn't called even though metadata was being set on the new hda. Am I missing a condition/something to trip this?

**Comment by natefoo** [comment]
@carlfeberhard I think you can remove it.

**Comment by carlfeberhard** [comment]
@natefoo thanks - will do possibly
